stl/inc/random

```diff
@@ -829,7 +829,7 @@ struct _Swc_traits { // traits for subtract_with_carry generator
 
         for (int _Ix = 0; _Ix < _Nw; ++_Ix) { // pack _Kx words
             _Ax[_Ix] = _Gx();
-            for (int _Jx = 0; ++_Jx < _Kx;) {
+            for (int _Jx = 1; _Jx < _Kx; ++_Jx) {

```

Comment:
This all looks fine, but the test failures (C6294 "Ill-defined for-loop:  initial condition does not satisfy test") make it clear why these loops are written this way in the first place - yet another static analysis rule that makes no sense to apply to value-dependent expressions. I'm inclined to suggest adding C6294 to the STL's list of globally ignored warnings in `yvals_core.h`, but let's see of anyone has different opinions before actually doing so.

Reply:
This loop fills `_Ax` array: outer part fills elements, and inner part fills 32-bit parts of each element.
Filling lowest 32-bits is moved out of the loop, see the line above, this operation also nullifies high bits.
Elements of `_Ax` array are of some integer type: see that `_Ty` is template value parameter.
So, as integers are currently limited to 64 bits, the loop will either execute zero times (for 8,16 or 32bit integers) or one time (for 32-bit integers).

The inner loop could have been unrolled like this:
```
if (_Kx == 0) {
_Ax[_Ix] = _Gx();
} else {
_Ax[_Ix] = _Gx() | (static_cast<_Ty>(_Gx()) << 32);
}
```
But then it stops being generic.
On the other hand, the code assumes 32 bit integers and 8 bit bytes, so it can hardly be called generic.

I'm skeptical on that changing the loop form is a significant improvement of clarity.

I do see a possible way to avoid warning suppression in this place `_Get_wc()` is effectively a `constexpr`, and can be marked as constexpr, then  loop entry can be guarded with constexpr if. Adding constexpr will also improve loop avoidance
```
static_assert(_Kx == 0 || _Kx == 1, "Wow, over 64 bit integers!");
if _CONSTEXPR_IF (_Kx == 0) {
```

Regarding the warning. Since the warning leads to the analysis above, it might not be totally useless.

Aside, the wording of the warning is confusing: _Ill-defined_ in C++ usually refers to undefined behavior, as opposed to _ill-formed_ which refers to errors that Standard require to be compile-time diagnosable. A loop that does not execute is not _Ill-defined_ by this definition.

Reply:
I have some newb questions, so please bear with me:
1. The original loop wouldn't have run after the test condition for 32 bit integers or lower, either (`++Jx` preincrements `_Jx` before comparing with `_Kx`). Why is it an error now, and wasn't already?
2. What about running the loop body for `_Jx == 0` as well? In that iteration, the loop body decomposes to: `_Ax[_Ix] |= static_cast<_Ty>(_Gx()) << (0);`. Starting with `_Ax[_Ix] = 0` takes care of garbage. Apart from the cost of an additional `static_cast`, I don't see the harm.

Reply:
I think that code has exactly the same semantic, except that the old one was not triggering static analysis error. The difference that make static analysis error show up is that after the change condition statement is an expression without side effects. Basically, the change served its purpose and made the code more clear for static analysis.

For your suggestion to zero initialize. Yes there are potentially additional costs. Though I wouldn't say that they are caused by static cast. Anyway, I'd expect these costs to optimize away. But if they wouldn't optimize away, then it is not a good change: random is on the critical path for significant amount of applications.

`_Kx * _Rx` produces compile time constant. Such constant expressions worked eve in pre C++11. Moreover, if you replace constants with preprocessor defines, it would work in C by defining usual array, _not_ VLA

Reply:
Here's a proof that changing loop initializer from `0` to `1` does not make difference, but zero-initializing before the loop or unrolling the loop makes difference: https://godbolt.org/z/kZPFnS

Reply:
Silencing C6294 "Ill-defined for-loop: initial condition does not satisfy test" STL-wide sounds like the right solution to me - I have never seen any problematic code identified by this warning, and it has caused trouble like this in the past.

Reply:
I guess I'll add the suppression in `yvals_core.h`.